main/src/library/BufReader.flix

```diff
@@ -0,0 +1,180 @@
+        rc: Region[r],
+        capacity: Int32,
+        reader: t
+    ): BufReader[t, r] \ r =
+        new BufReader @ rc {
+            rc = rc,
+            buffer = Array.repeat(rc, capacity, 0i8),
+            size = 0,
+            reader = reader
+        }
+
+    ///
+    /// Reads `k` bytes from the underlying `Reader` into `dest`, caching bytes read into
+    /// the internal buffer.
+    ///
+    /// Returns `Ok(k)` to signify that `k` bytes were successfully read and written to `dest`.
+    ///
+    /// Guarantees that `0 <= k <= length(dest)`.
+    ///
+    /// Returns `Err(e)` if some underlying error occurs.
+    ///
+    pub def read(
+        dest: Array[Int8, rb],
+        bufReader: BufReader[t, rr]
+    ): Result[IoError, Int32] \ rb + rr + Reader.Aef[t] with Reader[t] =
+        readFromCursor(0, dest, bufReader)
+
+    ///
+    /// Like `read`, but starting at a `cursor` which indicates the next index in `dest` to be written.
+    ///
+    /// Returns `Ok(c)` where `c` is a cursor indicating the next index in `dest` to be written.
+    ///
+    /// Guarantees that `0 <= c <= length(dest)`.
+    ///
+    /// The number of bytes read is `c - cursor`, (which is just `c` when `cursor` is 0).
+    ///
+    /// Assumes that `cursor <- length(dest)`.
+    ///
+    def readFromCursor(
+        cursor: Int32,
+        dest: Array[Int8, rb],
+        bufReader: BufReader[t, rr]
+    ): Result[IoError, Int32] \ rb + rr + Reader.Aef[t] with Reader[t] =
+        if (cursor == Array.length(dest)) {
+            Ok(cursor)
+        } else if (bufReader->size != 0) {
+            let bytesCopied = copyFromBuffer(cursor, dest, bufReader);
+            readFromCursor(cursor + bytesCopied, dest, bufReader)
+        } else {
+            forM(
+                bytesFilled <- fillBuffer(bufReader);
+                newCursor <- if (bytesFilled == 0) {
+                    Ok(cursor)
+                } else {
+                    readFromCursor(cursor, dest, bufReader)
+                }
+            ) yield newCursor
+        }
+
+    ///
+    /// Copies from the buffer into the destination, removing copied bytes from the buffer.
+    ///
+    /// Does not read from the underlying reader.
+    ///
+    /// Returns the number of bytes copied.
+    ///
+    /// Assumes that `cursor < length(dest)`.
+    ///
+    def copyFromBuffer(
+        cursor: Int32,
+        dest: Array[Int8, rb],
+        bufReader: BufReader[t, rr]
+    ): Int32 \ rb + rr with Reader[t] =
+        let bytesToRead = Int32.min(Array.length(dest) - cursor, bufReader->size);
+        Array.copyInto(
+            srcPos = 0,
+            dstPos = cursor,
+            len = bytesToRead,
+            src = bufReader->buffer,
+            dest
+        );
+        shiftBufferLeft(bytesToRead, bufReader->buffer);
+        bufReader->size = bufReader->size - bytesToRead;
+        bytesToRead
+
+    def shiftBufferLeft(n: Int32, buffer: Array[a, r]): Unit \ r =
```

Comment:
Would it make sense to put this in the array module

Reply:
I think that's a good idea if you rename it to `rotateInPlace`. Also, the implementation should probably just use a local visitor that increments `i` every recursive call instead of using `foreach` :)